fulltext/controllers.py

```diff
@@ -1,105 +1,239 @@
-        }, HTTP_500_INTERNAL_SERVER_ERROR
-    except Exception as e:
-        return {'explanation': str(e)}, HTTP_500_INTERNAL_SERVER_ERROR
-    if content_data is None:
-        return {
-            'explanation': 'fulltext not available for %s' % document_id
-        }, HTTP_404_NOT_FOUND
-    return content_data, HTTP_200_OK
-
-
-def extract(payload: str) -> tuple:
-    """Handle a request for reference extraction."""
-    document_id = payload.get('document_id')
-    if document_id is None or not isinstance(document_id, str):
-        return DOCUMENT_ID_MISSING, HTTP_400_BAD_REQUEST, {}
-    logger.info('extract: got document_id: %s' % document_id)
-
-    latest = store.latest(document_id)
-    if latest is not None and latest.get('version') >= store.current_version():
-        headers = {
-            'Location': url_for('fulltext.retrieve', doc_id=document_id)
-        }
-        return ALREADY_EXISTS, HTTP_303_SEE_OTHER, headers
-    logger.info('extract: got latest: %s' % latest)
-
-    pdf_url = payload.get('url')
-    if pdf_url is None or not retrievePDF.is_valid_url(pdf_url):
-        return FILE_MISSING_OR_INVALID, HTTP_400_BAD_REQUEST, {}
-    logger.info('extract: got url: %s' % pdf_url)
-
-    result = extract_fulltext.delay(document_id, pdf_url)
-    logger.info('extract: started processing as %s' % result.task_id)
-    headers = {'Location': url_for('fulltext.task_status',
-                                   task_id=result.task_id)}
-    return ACCEPTED, HTTP_202_ACCEPTED, headers
-
-
-def status(task_id: str) -> tuple:
+        raise InternalServerError('Could not connect to backend') from e
+    except store.DoesNotExist:
+        # If there is only metadata, we should still get a response from
+        # the store. So if we hit DoesNotExist there really is nothing to see
+        # here folks, move along.
+        raise NotFound('No such extraction')
+
+    # Make sure that the client is authorized to work with this resource before
+    # doing anything else.
+    if authorizer and not authorizer(identifier, product.owner):
+        raise NotFound('No such extraction')
+
+    if product.content is None \
+            and product.status is Extraction.Status.IN_PROGRESS:
+        target = url_for('fulltext.task_status', identifier=identifier,
+                         id_type=id_type)
+        return TASK_IN_PROGRESS, status.SEE_OTHER, {'Location': target}
+    return product.to_dict(), status.OK, {}
+
+
+def start_extraction(id_type: str, identifier: str, token: str,
+                     force: bool = False,
+                     authorizer: Optional[Authorizer] = None) -> Response:
+    """Handle a request to force text extraction."""
+    if id_type not in SupportedBuckets:
+        raise NotFound('Unsupported identifier')
+
+    canonical = pdf.CanonicalPDF.current_session()
+    storage = store.Storage.current_session()
+    compilations = compiler.Compiler.current_session()
+
+    if not force:
```

Comment:
The non-force case does not check if article exists before redirecting to fulltext. 

The full text should probably not be available if there is no article. Ex. a takedown of an article.

Reply:
Good catch. Fixed in ec36e33